clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java

```diff
@@ -239,66 +246,106 @@ public void fetch() {
+            } else if (!acknowledgeRequestState.retryTimeoutExpired(currentTimeMs)) {
+                if (acknowledgeRequestState.canSendRequest(currentTimeMs)) {
+                    acknowledgeRequestState.onSendAttempt(currentTimeMs);
+                    unsentRequests.add(acknowledgeRequestState.buildRequestOnClose(currentTimeMs));
+                }
+            } else {
+                // Fill in TimeoutException
+                for (TopicIdPartition tip : acknowledgeRequestState.acknowledgementsMap.keySet()) {
+                    metricsManager.recordFailedAcknowledgements(shareFetchBuffer.getPendingAcknowledgementsCount(tip));
+                    shareFetchBuffer.handleAcknowledgementResponses(tip, Errors.REQUEST_TIMED_OUT);
+                }
+                acknowledgeRequestStates.remove(acknowledgeRequestState);
+            }
+        }
+        PollResult pollResult = null;
+        if (!unsentRequests.isEmpty()) {
+            pollResult = new PollResult(unsentRequests);
+        } else if (acknowledgeRequestStates.isEmpty()) {
+            // Complete the future.
+            if (commitSyncFuture != null) {
+                commitSyncFuture.complete(null);
+            }
+        } else {
+            // Return empty result until all the acknowledgement request states are processed.
+            pollResult = PollResult.EMPTY;
+        }
+        return pollResult;
+    }
+
+    private PollResult processAcknowledgements(long currentTimeMs) {
+        List<UnsentRequest> unsentRequests = new ArrayList<>();
+        for (AcknowledgeRequestState acknowledgeRequestState : acknowledgeRequestStates) {
+            if (acknowledgeRequestState.isExpired()) {
+                acknowledgeRequestStates.remove(acknowledgeRequestState);
+            } else if (!acknowledgeRequestState.retryTimeoutExpired(currentTimeMs)) {
+                if (acknowledgeRequestState.canSendRequest(currentTimeMs)) {
+                    acknowledgeRequestState.onSendAttempt(currentTimeMs);
+                    unsentRequests.add(acknowledgeRequestState.buildRequest(currentTimeMs));
+                }
+            } else {
+                // Fill in TimeoutException
+                for (TopicIdPartition tip : acknowledgeRequestState.acknowledgementsMap.keySet()) {
+                    metricsManager.recordFailedAcknowledgements(shareFetchBuffer.getPendingAcknowledgementsCount(tip));
+                    shareFetchBuffer.handleAcknowledgementResponses(tip, Errors.REQUEST_TIMED_OUT);
+                }
+                acknowledgeRequestStates.remove(acknowledgeRequestState);
+            }
+        }
+        PollResult pollResult = null;
+        if (!unsentRequests.isEmpty()) {
+            pollResult = new PollResult(unsentRequests);
+        } else if (acknowledgeRequestStates.isEmpty()) {
+            // Complete the future.
+            if (commitSyncFuture != null) {
+                commitSyncFuture.complete(null);
+            }
```

Comment:
Maybe you should assign `commitSyncFuture = null` before you complete it since that's logically the end of the operation.